### PR TITLE
Train rework 2

### LINKED ISSUE #77: checkpoint parameterization
Train checkpointing should be parameterized on 'every n steps' as well as a specific (user-supplied) list of steps to checkpoint on (main use case is high frequency checkpoint early in training to study how the model develops early on)

checkpointing should *never* be a function of validation performance, should be based entirely on configuration; you should know exactly how many checkpoints will be created before training starts, and where they will be

### LINKED ISSUE #83: resume_from_dir should be separate argument
distinct fro output_dir

### LINKED ISSUE #84: huggingface upload logic
Checkpoints should be saved to their own branches on huggingface repos, should not include optimizer weights

### PR DESCRIPTION
**Pull request by jaidhyani**

### CODE REVIEW

File: scripts/training_config_examples/sample_transformers_bloom.json [hunk 1]
```diff
@@ -7,7 +7,7 @@
     "eval_iters": 8,
     "batch_size": 64,
     "model_config": {
-        "model_type": "BloomForCausalLM",
+        "model_class": "BloomForCausalLM",
         "model_params": {
```
Comment (jettjaniak) [inline scripts/training_config_examples/sample_transformers_bloom.json]: that is wrong

File: src/delphi/train/checkpoint_step.py [hunk 2]
```diff
@@ -1,28 +1,26 @@
...
 
 
 def should_run_checkpoint(config: GigaConfig, mts: ModelTrainingState):
-    return mts.iter_num % config.eval_interval == 0
+    return any(
+        [
+            mts.iter_num % config.checkpoint_interval == 0 and mts.iter_num > 0,
+            mts.iter_num in config.extra_checkpoint_iters,
+        ]
+    )
```
Comment (jettjaniak) [inline src/delphi/train/checkpoint_step.py]: or

File: src/delphi/train/checkpoint_step.py [hunk 3]
```diff
@@ -1,28 +1,26 @@
...
-    CheckpointData,
-    ModelTrainingState,
-    estimate_loss,
-    save_checkpoint_if_needed,
-)
+from .utils import ModelTrainingState, count_tokens_so_far, estimate_loss, save_results
 from .wandb_utils import log_to_wandb
 
 
 def should_run_checkpoint(config: GigaConfig, mts: ModelTrainingState):
```
Comment (jettjaniak) [inline src/delphi/train/checkpoint_step.py]: bool type hint

File: src/delphi/train/config/gigaconfig.py [hunk 4]
```diff
@@ -1,63 +1,74 @@
...
     )
 
-    # I/O
-    eval_interval: int = field(default=2000, metadata={"help": "eval every N iters"})
+    # checkpoints, logging, eval
+    checkpoint_interval: int = field(
+        default=2000, metadata={"help": "checkpoint every N iters"}
+    )
+    extra_checkpoint_iters: list[int] = field(
+        default_factory=lambda: [],
```
Comment (jettjaniak) [inline src/delphi/train/config/gigaconfig.py]:
```suggestion
        default_factory=list,
```

File: src/delphi/train/config/gigaconfig.py [hunk 5]
```diff
@@ -1,63 +1,74 @@
...
-        metadata={"help": "if True, always save a checkpoint after each eval"},
-    )
+
+    # initialization
     init_from: str = field(
         default="scratch",
         metadata={
             "help": "'scratch' for a new model, 'resume' to resume from output_dir"
         },
     )
```
Comment (jettjaniak) [inline src/delphi/train/config/gigaconfig.py]: drop it

File: src/delphi/train/config/gigaconfig.py [hunk 6]
```diff
@@ -1,63 +1,74 @@
...
-    # wandb logging
-    wandb_config: WandbConfig = field(default_factory=WandbConfig)
+    resume_from_path: str = field(
+        default=".",
```
Comment (jettjaniak) [inline src/delphi/train/config/gigaconfig.py]:
```suggestion
        default=None,
```

File: src/delphi/train/config/optimizer_config.py [hunk 7]
```diff
@@ -3,13 +3,13 @@
 
 @dataclass
 class OptimizerConfig:
```
Comment (jettjaniak) [inline src/delphi/train/config/optimizer_config.py]:
```suggestion
class AdamConfig:
```

File: src/delphi/train/utils.py [hunk 8]
```diff
@@ -113,32 +101,11 @@ def set_lr(
...
-        results_path=results_path,
-    )
-
-
 def initialize_model_training_state(
     config: GigaConfig, device: torch.device
 ) -> ModelTrainingState:
     t0 = time.time()
-    model = config.model_config.get_model()
+    model = get_model(config.model_config)
```
Comment (jettjaniak) [inline src/delphi/train/utils.py]:
```suggestion
    model = init_model(config.model_config)
```
maybe?

Comment (jettjaniak) [reply src/delphi/train/utils.py]: maybe take seed and re-seed?

**Review by jettjaniak — APPROVED** [review]
🚀

**Comment by jaidhyani** [comment]
see #95

**Comment by jaidhyani** [comment]
> 🚀

Could you copy this approval to #95?

**Pull request closed without merging**